### PR TITLE
Add tribool type and update is_zero to return tribool

### PR DESCRIPTION
**Pull request by rikardn**
This is a reimplementation of the `is_zero` function. It can now return a `tribool` with an additional indeterminate value. It is implemented using a visitor, which so far doesn't have much more functionality than before, but has the potential. 

Future extension could be:
* Enable the visitor to use a context i.e. assumptions
* Other tests like `is_positive` and `is_real`
* More advanced zero-testing, e.g. for `mul` to check if any factor is zero etc

### CODE REVIEW

**Comment by isuruf** [comment]
Thanks for sending the PR.

I'd rather not do this in critical code paths inside the library as the visitor can be expensive and adding future extensions to this is going to make the function even slower. Maybe we can have two functions?

**Comment by rikardn** [comment]
Ok, agreed. How should I name the functions? Let the old faster function be called `is_number_and_zero` ?

**Comment by isuruf** [comment]
`is_number_and_zero` sounds good. For matrix functions we can keep using `is_zero` where pivoting decisions need the best `is_zero` we can have.

**Comment by rikardn** [comment]
So now the previous `is_zero` is back as `is_number_and_zero` and this function is used in pow.cpp and mul.cpp

File: symengine/test_visitors.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    };
+    void bvisit(const Constant &x)
+    {
+        is_zero_ = tribool::trifalse;
+    };
+
+    tribool apply(const Basic &b);
+};
+
+tribool is_zero(const Basic &b);
```
Comment (isuruf) [inline symengine/test_visitors.h:50]: Can you move this declaration to `number.h`?

**Review by isuruf — COMMENTED** [review]
One comment, otherwise looks good

**Comment by rikardn** [comment]
Ok, I have moved the `is_zero` declaration to number.h. I had to also move the `tribool` declaration there to avoid circular includes (vistor.h includes number.h). Now it feels like `tribool` is a bit out of place. Perhaps it should have its own header or be in a top level symengine.h or similar.

**Comment by isuruf** [comment]
Yes, `tribool` definition and `is_zero` definition can be moved to `basic.h`.

**Comment by isuruf** [comment]
I meant, `is_true`

**Comment by rikardn** [comment]
Agree! I have moved `is_true` and `tribool` to `basic.h`

**Comment by isuruf** [comment]
Thanks @rikardn

**Pull request merged**